mt-kahypar/partition/refinement/flows/parallel_construction.cpp

```diff
@@ -304,12 +311,192 @@ FlowProblem ParallelConstruction<GraphAndGainTypes>::constructDefault(const Part
+      FlowNetworkConstruction::getParameters(phg, _context, he, block_0, block_1);
+    ASSERT(parameters.capacity >= 0);
+
+    // mark HE as invalid if it is not later explicitly marked as valid
+    valid_hyperedges[i] = static_cast<uint8_t>(false);
+
+    // this check removes any edges that can't affect the global objective
+    if (parameters.capacity == 0) return;
+
+    vec<whfc::Node>& tmp_pins = tmp_hyperedges[i].pins;
+    // Copy hyperedge and pins to temporary buffer
+    size_t he_hash = 0;
+    tmp_hyperedges[i].weight = parameters.capacity;
+    tmp_pins.clear();
+    tmp_pins.reserve(phg.edgeSize(he));
+    if ( parameters.is_cut || (phg.pinCountInPart(he, block_0) > 0 && phg.pinCountInPart(he, block_1) > 0) ) {
+      __atomic_fetch_add(&flow_problem.total_cut, parameters.capacity, __ATOMIC_RELAXED);
+    }
+    for (const HypernodeID& pin : phg.pins(he)) {
+      whfc::Node* whfc_pin = _node_to_whfc.get_if_contained(pin);
+      if (whfc_pin) {
+        push_into_tmp_pins(tmp_pins, *whfc_pin, he_hash, false);
+      } else {
+        const PartitionID pin_block = phg.partID(pin);
+        parameters.connect_to_source |= pin_block == block_0;
+        parameters.connect_to_sink |= pin_block == block_1;
+      }
+    }
+
+    const bool empty_hyperedge = tmp_pins.size() == 0;
+    const bool connected_to_source_and_sink = parameters.connect_to_source && parameters.connect_to_sink;
+    if (connected_to_source_and_sink) {
+      // Hyperedge is connected to source and sink which means we can not remove it
+      // from the cut with the current flow problem => remove he from flow problem
+      __atomic_fetch_add(&flow_problem.non_removable_cut, tmp_hyperedges[i].weight, __ATOMIC_RELAXED);
+    } else if (!empty_hyperedge) {
+      if (parameters.connect_to_source) {
+        push_into_tmp_pins(tmp_pins, flow_problem.source, he_hash, true);
+      } else if (parameters.connect_to_sink) {
+        push_into_tmp_pins(tmp_pins, flow_problem.sink, he_hash, true);
+      }
+
+      if (tmp_pins.size() > 1) {
+        // Insert contracted hyperedge
+        hyperedge_hash_map.insert(he_hash,
+          HashedHyperedgeInformation{ i, he_hash, tmp_pins.size() });
+        valid_hyperedges[i] = static_cast<uint8_t>(true);
+        std::sort( tmp_pins.begin() +
+                  ( tmp_pins[0] == flow_problem.source ||
+                    tmp_pins[0] == flow_problem.sink), tmp_pins.end());
+      }
+    }
+  });
+
+  tbb::parallel_for(UL(0), hyperedge_hash_map.numBuckets(), [&](const size_t bucket) {
+    auto& hyperedge_bucket = hyperedge_hash_map.getBucket(bucket);
+    std::sort(hyperedge_bucket.begin(), hyperedge_bucket.end(),
+      [&](const HashedHyperedgeInformation& lhs, const HashedHyperedgeInformation& rhs) {
+      return std::tie(lhs.hash, lhs.size, lhs.he) < std::tie(rhs.hash, rhs.size, rhs.he);
+    });
+
+    // Parallel Hyperedge Detection
+    for (size_t i = 0; i < hyperedge_bucket.size();) {
```

Comment:
This is a bit different from the usual INR algorithm since it makes a single forward pass through the hyperedges. Say we have a four HEs |e0|e1|e2|e3| with equal hashes and e0 == e2, e1 == e3. Then the first pair (e0, e2) is found, but (e1, e3) is not. Whether this is fine or not is up for debate I suppose. INR is just a performance optimization, not necessary for correctness. Whether we spend quadratic work here, or do some redundant work in the flow computation, it's unclear which is favorable. Though I would tend to prefer doing quadratic work here. If it turns out to be problematic, we can look into performance improvements again.

Reply:
Yeah, makes sense